ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -87,41 +89,83 @@ private[platform] final class EventsBuffer[O: Ordering, E](
     * @param endInclusive The end inclusive bound of the requested range.
     * @return The series of events as an ordered vector satisfying the input bounds.
     */
-  def slice(startExclusive: O, endInclusive: O): BufferSlice[(O, E)] =
+  def slice[FILTER_RESULT](
+      startExclusive: Offset,
+      endInclusive: Offset,
+      filter: ENTRY => Option[FILTER_RESULT],
+  ): BufferSlice[(Offset, FILTER_RESULT)] =
     Timed.value(
       sliceTimer, {
         val bufferSnapshot = _bufferStateRef
+        val vectorSnapshot = bufferSnapshot.vector
+
         if (bufferSnapshot.rangeEnd.exists(_ < endInclusive)) {
-          throw RequestOffBufferBounds(bufferSnapshot.vector.last._1, endInclusive)
-        } else if (bufferSnapshot.vector.isEmpty)
-          Empty
-        else {
-          val Seq(bufferStartInclusiveIdx, bufferEndExclusiveIdx) =
-            Seq(startExclusive, endInclusive)
-              .map(bufferSnapshot.vector.searchBy(_, _._1))
-              .map {
-                case InsertionPoint(insertionPoint) => insertionPoint
-                case Found(foundIndex) => foundIndex + 1
-              }
-
-          val vectorSlice =
-            bufferSnapshot.vector.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
-
-          if (bufferStartInclusiveIdx == 0) Prefix(vectorSlice)
-          else Inclusive(vectorSlice)
+          throw RequestOffBufferBounds(bufferSnapshot.rangeEnd.get, endInclusive)
+        } else {
+          val bufferStartInclusiveIdx = searchByOffset(vectorSnapshot, startExclusive)
+          val bufferEndExclusiveIdx = searchByOffset(vectorSnapshot, endInclusive)
+          val bufferSlice = vectorSnapshot.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
+
+          if (bufferStartInclusiveIdx == 0) {
+            // The buffer is potentially a suffix of the fetched window
+            if (bufferSlice.isEmpty) {
+              // If the slice is empty, the entire window is before the buffer bounds
+              Suffix(endInclusive, Vector.empty, None)
+            } else {
+              val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice.tail, filter)
+              // We "waste" the first element of the slice in order to provide
```

Comment:
I guess we could be smarter about this is we considered if the `buffer.searchBy` returned `Found`.  Only possibly worth considering from a latency perspective in the case of a single transaction being played in ?

Reply:
What I was trying to say was that if the full range was found in the buffer, starting at position 0, then maybe an `Inclusive` could be returned instead of a `Suffix` (so no need for `endInclusive`) ?

Reply:
This is the current behavior. if the index is found in the buffer at `Found(0)`, `bufferStartInclusiveIdx` is `1` which takes the `Inclusive(..)` branch

Reply:
I understand it was not trivial to follow what's going on. Tried a refactoring for better visibility: bfba6d372cf7ff2cf3757c6108a00c6719ea0666

Reply:
The changes make it much clearer what is going on, thanks.